src/bin/playground.cpp

```diff
@@ -1,10 +1,278 @@
+   * As a next step we should use the AttributeVectorCompressionID to define the type of the FixedWidthIntegerVector
+   * and perhaps also write out the type of the DictionarySegment.
+   */
+
+  std::ofstream chunk_file;
+  chunk_file.open(filename, std::ios::out | std::ios::binary | std::ios::app);
+
+  // We will later mmap to an uint32_t vector/array. Therefore, we store all metadata points as uint32_t.
+  // This wastes up to three bytes of compression per metadata point but makes mapping much easier.
+  export_value(chunk_file, static_cast<uint32_t>(segment->dictionary()->size()));
+  export_value(chunk_file, static_cast<uint32_t>(segment->attribute_vector()->size()));
+
+  const auto compressed_vector_type_id = static_cast<uint32_t>(infer_compressed_vector_type_id<int>(*segment));
+  export_value(chunk_file, compressed_vector_type_id);
+
+  export_values<int>(chunk_file, *segment->dictionary());
+
+  export_compressed_vector(chunk_file, *segment->compressed_vector_type(),
+                          *segment->attribute_vector());
+  chunk_file.close();
+}
+
+void write_chunk(const dict_chunk_prototype& chunk, const std::string& chunk_filename) {
+  const auto file_extension = ".bin";
+  const auto filename = chunk_filename + file_extension;
+  const auto segment_count = chunk.size();
+  for (auto segment_index = size_t{0}; segment_index < segment_count; ++segment_index) {
+    write_dict_segment(chunk[segment_index], filename);
+  }
+}
+
+dict_chunk_prototype map_chunk(const std::string& chunk_name, const uint32_t segment_count) {
+  auto mapped_dict_chunk = dict_chunk_prototype{};
+
+  auto fd = int32_t{};
+  const auto file_extension = ".bin";
+  const auto chunk_filename = chunk_name + file_extension;
+  Assert((fd = open(chunk_filename.c_str(), O_RDONLY)) >= 0, fail_and_close_file(fd, "Open error: ", errno));
+
+  const auto offset = off_t{0};
+
+  // As we store a variable number of bytes per segment the easiest solution is to
+  // obtain the bytes to mmap via a file system call.
+  const auto file_bytes = std::filesystem::file_size(chunk_filename);
+
+  auto* map = reinterpret_cast<uint32_t*>(mmap(NULL, file_bytes, PROT_READ, MAP_PRIVATE, fd, offset));
+  Assert((map != MAP_FAILED), fail_and_close_file(fd, "Mapping Failed: ", errno));
+  close(fd);
+
+  madvise(map, file_bytes, MADV_SEQUENTIAL);
+
+  auto currently_mapped_elements = uint32_t{0};
+  const auto meta_data_elements = uint32_t{3};
+  for (auto segment_index = size_t{0}; segment_index < segment_count; ++segment_index) {
+    const auto dictionary_size = map[currently_mapped_elements];
+    const auto attribute_vector_size = map[currently_mapped_elements + 1];
+    //const auto encoding_type = map[currently_mapped_elements + 2]; //currently unused, see `write_dict_chunk` comment
+
+    // As a first step we don't use the mmap as the underlying data structure of the DictionarySegment but
```

Comment:
We currently don't use/access the mmapped memory directly. This is opposed to our exercise example with ValueSegment-ish datastructures, where we simply wrapped a std::span around the mmapped memory and used it for vector/datastructure equivalent access. This is not possible here as we use actual DictionarySegments that require a pmr_vector or a BaseCompressedVector as data structure for their dictionary and attribute_vector, respectively.
As far as I know, changing the pointer of the underlying data of a vector (which can be accessed via vector.data()) is not possible as it opposes the idea of the vector self-managing its storage. I assume that we will need to adapt DictionarySegment to also accept std::span datastructures or perhaps write a DiskDictionarySegment class that handles std::spans?

(Inspired by the pmr_vector I shortly looked into registering the mmap as memory resource for the DictionarySegment but that still doesn't solve the self-managing storage issue / doesn't initialize with the values from the mmap. (And is intended for completely different use-cases I assume.)

Reply:
You are exactly right. There is no "clean" way to use std::vectors in this case (to my best knowledge). What we thought about is adapting the dictionary segment to use std::spans.
Than could be done in several steps to keep the changes low. (i) Have internal std::unique_ptrs to "some memory" and use a  std::spans on top of these (instead of our current vectors). (ii) move the memory management to the storage manager. (iii) enable to the storage manager to use an mmap instead of the DRAM allocation.

This is just an example. In general the idea is to evaluate how easy we can adapt dictionary and value segments in Hyrise. With that change, the storage manager will probably finally life up to its name and actually manage "the storage" (someone needs to keep track of large mmap's and so on).

Reply:
Hi, thank you for the feedback/answer, that makes a lot of sense.